### ISSUE TITLE
Use non-blocking CSS

### ISSUE THREAD

**Issue by matthewp** [donejs/donejs#679]
Currently donejs does something which goes against [PageSpeed](https://developers.google.com/speed/pagespeed/) recommendations and inserts all of the css into the head as `<link>`s. This will block the page from rendering as these stylesheets are loaded.

We should change this and do a couple of things:
1. Inline "essential" styles directly in `<style>` tags. This might just be the main css bundle. Ideally this could even be trimmed down to just selectors that are being used by the page.
2. Make the rest of the links be non-blocking by using something like [loadCSS](https://github.com/filamentgroup/loadCSS#recommended-usage-pattern).

**Comment by justinbmeyer** [comment]
If those styles are going to be used on other pages (that would not be reached via push state), I'm not sure #1 is good advice. 

#2 doesn't make sense with server side rendering. You need CSS to load before people see HTML. 

I think the best thing we could do is send the head/link tags out asap, and stream the rest later.

**Comment by matthewp** [comment]
> If those styles are going to be used on other pages (that would not be reached via push state), I'm not sure #1 is good advice.

That's true, you need a size threshold to determine whether it is worth it. Small css can be inlined but larger should not be.

I think you still want to asynchronously load the CSS even if it is a larger file that becomes a link. Unstyled content is better than no content.  But this could be configurable perhaps.

**Comment by matthewp** [comment]
[This tool](https://github.com/addyosmani/critical) can be used to extract above-the-fold CSS which we could use to attach to the `<head>` of the page:

``` html
<head>
  <style>
    . body {
      /* above the fold css */
    }
  </style>

  <link rel="preload" href="/dist/bundles/below-the-fold.css" as="style" onload="this.rel='stylesheet'">
  <link rel="preload" href="/dist/bundles/even-more.css" as="style" onload="this.rel='stylesheet'">
</head>
```

**Comment by DesignByOnyx** [comment]
I have a lot of experience doing this and getting a good rapport with PageSpeed is very difficult. Inlining critical CSS and deferring other CSS is really the only way. Pretty much anything that requires an http request before the DOMContentLoaded event fires (CSS `<link>`s and synchronous `<script>` tags) will cost you favor with PageSpeed. It basically wants your initial HTML payload to deliver everything required to render something for the user without waiting on extra requests.

**Comment by frank-dspeed** [comment]
i think its realy a good idea to add loadCSS Functionality but we should add a wrapper around it that accepts config like inline defer for each css that it loads also the wrapper should be pushstate agnostic so that it still allows reaching inlined and not inlined css files
